packages/typespec-client-generator-core/test/decorators/inherits-from.test.ts

```diff
@@ -0,0 +1,375 @@
+import { expectDiagnostics } from "@typespec/compiler/testing";
+import { ok, strictEqual } from "assert";
+import { beforeEach, it } from "vitest";
+import { SdkTestRunner, createSdkTestRunner } from "../test-host.js";
+
+let runner: SdkTestRunner;
+
+beforeEach(async () => {
+  runner = await createSdkTestRunner({ emitterName: "@azure-tools/typespec-java" });
+});
+
+it("three-level inheritance chain", async () => {
+  await runner.compileWithBuiltInService(`
+      @discriminator("kind")
+      model A {
+        kind: string;
+      }
+
+      model BContent {
+        foo: string;
+      }
+
+      model B extends A {
+        kind: "B";
+        ...BContent;
+      }
+
+      @inheritsFrom(B)
+      model C extends A {
+        kind: "C";
+        ...BContent;
+        bar: string;
+      }
+
+      @route("/test")
+      op test(): A;
+    `);
+
+  const models = runner.context.sdkPackage.models;
+  const modelA = models.find((m) => m.name === "A");
+  const modelB = models.find((m) => m.name === "B");
+  const modelC = models.find((m) => m.name === "C");
+
+  ok(modelA);
+  ok(modelB);
+  ok(modelC);
+
+  // C should inherit from B instead of A due to @inheritsFrom
+  strictEqual(modelC.baseModel?.name, "B");
+  strictEqual(modelB.baseModel?.name, "A");
+
+  // Verify discriminator property is correctly identified
+  strictEqual(modelA.discriminatorProperty?.name, "kind");
+  strictEqual(modelB.discriminatorValue, "B");
+  strictEqual(modelC.discriminatorValue, "C");
+  strictEqual(modelA.discriminatorValue, undefined);
+
+  // Verify properties are correctly inherited
+  strictEqual(modelC.properties.length, 3);
+  strictEqual(modelC.properties[0].name, "kind");
+  strictEqual(modelC.properties[1].name, "foo");
+  strictEqual(modelC.properties[2].name, "bar");
```

Comment:
I think there is a logic missing here. `C` should not have the property of `B`: `foo`. Also, as @ArcturusZhang mentioned, I wonder if we should take this [solution](https://github.com/Azure/typespec-azure/issues/3015#issuecomment-3099326745).